common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,985 @@
+                continue;
+            }
+
+            ValueType h_k_i_scalar = h_k_i_shared_entry.values[c];
+
+            w_shared_entry.values[r * w_shared_entry.stride + c] -=
+                h_k_i_scalar *
+                v_k_shared_entry.values[r * v_k_shared_entry.stride + c];
+        }
+
+        __syncthreads();
+    }
+
+
+    __shared__
+        UninitializedArray<real_type, batch_config<ValueType>::max_num_rhs>
+            w_norms_sh;
+
+    gko::batch_dense::BatchEntry<real_type> w_norms_shared_entry{
+        w_norms_sh, w_shared_entry.stride, 1, w_shared_entry.num_rhs};
+
+    compute_norm2(gko::batch::to_const(w_shared_entry), w_norms_shared_entry);
+
+    const gko::batch_dense::BatchEntry<ValueType> h_i_plus_1_i_shared_entry{
+        &H_shared_entry.values[(i + 1) * H_shared_entry.stride + i * num_rhs],
+        static_cast<size_type>(w_shared_entry.num_rhs), 1,
+        w_shared_entry.num_rhs};
+
+    __syncthreads();
+
+
+    for (int c = threadIdx.x; c < w_norms_shared_entry.num_rhs;
+         c += blockDim.x) {
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        h_i_plus_1_i_shared_entry.values[c] =
+            static_cast<ValueType>(w_norms_shared_entry.values[c]);
+    }
+
+
+    __syncthreads();
+
+    const gko::batch_dense::BatchEntry<ValueType> v_i_plus_1_shared_entry{
+        &V_shared_entry.values[(i + 1) * V_shared_entry.stride * num_rows],
+        V_shared_entry.stride, num_rows, V_shared_entry.num_rhs};
+
+
+    for (int li = threadIdx.x;
+         li < w_shared_entry.num_rows * w_shared_entry.num_rhs;
+         li += blockDim.x) {
+        int r = li / w_shared_entry.num_rhs;
+        int c = li % w_shared_entry.num_rhs;
+
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        v_i_plus_1_shared_entry.values[r * v_i_plus_1_shared_entry.stride + c] =
+            w_shared_entry.values[r * w_shared_entry.stride + c] /
+            h_i_plus_1_i_shared_entry.values[c];
+    }
+}
+
+
+template <typename ValueType>
+__device__ __forceinline__ void GeneratePlaneRotation(
+    const ValueType *const x, const ValueType *const y, ValueType *const cs,
+    ValueType *const sn, const int nrhs, const uint32 &converged)
+{
+    for (int c = threadIdx.x; c < nrhs; c += blockDim.x) {
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        if (y[c] == zero<ValueType>()) {
+            cs[c] = one<ValueType>();
+            sn[c] = zero<ValueType>();
+        } else if (abs(y[c]) > abs(x[c])) {
+            ValueType temp = x[c] / y[c];
+            sn[c] = one<ValueType>() / sqrt(one<ValueType>() + temp * temp);
+            cs[c] = temp * sn[c];
+        } else {
+            ValueType temp = y[c] / x[c];
+            cs[c] = one<ValueType>() / sqrt(one<ValueType>() + temp * temp);
+            sn[c] = temp * cs[c];
```

Comment:
It would be nice if complex could be addressed, but I don't think it's urgent right now. We could come back to it later too. If that is the case, perhaps we should throw an exception in the apply_impl for complex instantiations.

Reply:
I have now tried adapting it for complex systems.